tests/queryparams/query_params_test.go

```diff
@@ -0,0 +1,232 @@
+		"Host: localhost:8080\r\n" +
+		"\r\n"
+	resp, err := requesttesting.MakeRequest(context.Background(), []byte(req), func(req *http.Request) {
+		want := url.Values{"vegetable": []string{"potatO", "Tomato"}}
+		got := req.URL.Query()
+		if diff := cmp.Diff(want, got); diff != "" {
+			t.Errorf("req.URL.Query(): got %v, want %v, diff (-want +got):\n%s", got, want, diff)
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(status200OK)) {
+		t.Errorf("response status: got %s, want %s", resp, status200OK)
+	}
+}
+
+func TestQueryParametersSameKeyDifferentCasing(t *testing.T) {
+	req := []byte("GET /?vegetable=potato&Vegetable=tomato HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		want := url.Values{
+			"vegetable": []string{"potato"},
+			"Vegetable": []string{"tomato"},
+		}
+		got := req.URL.Query()
+		if diff := cmp.Diff(want, got); diff != "" {
+			t.Errorf("req.URL.Query(): got %v, want %v, diff (-want +got): \n%s", got, want, diff)
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(status200OK)) {
+		t.Errorf("response status: got %s, want %s", resp, status200OK)
+	}
+}
+
+// TestQueryParametersValidUnicode ensures keys and values that contain non-ASCII characters are parsed correctly
+func TestQueryParametersValidUnicode(t *testing.T) {
+	req := []byte("GET /?vegetable=ăȚâȘî HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		want := url.Values{"vegetable": []string{"ăȚâȘî"}}
+		got := req.URL.Query()
+		if diff := cmp.Diff(want, got); diff != "" {
+			t.Errorf("req.URL.Query(): got %v, want %v, diff (-want +got):\n%s", got, want, diff)
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(status200OK)) {
+		t.Errorf("response status: got %s, want %s", resp, status200OK)
+	}
+
+	req = []byte("GET /?ăȚâȘî=vegetable HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"\r\n")
+	resp, err = requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		want := url.Values{"ăȚâȘî": []string{"vegetable"}}
+		got := req.URL.Query()
+		if diff := cmp.Diff(want, got); diff != "" {
+			t.Errorf("req.URL.Query(): got %v, want %v, diff (-want +got):\n%s", got, want, diff)
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(status200OK)) {
+		t.Errorf("response status: got %s, want %s", resp, status200OK)
+	}
+}
+
+// Tests whether passing invalid Unicode in both key and value will result in a 400 Bad Request response
+func TestQueryParametersInvalidUnicodes(t *testing.T) {
+	req := []byte("GET /?\x0F=tomato&Vegetable=potato HTTP/1.1\r\n" +
```

Comment:
nit: tabletest?

Reply:
Missed it, thanks! Resolved in [9a7e610](url)